### PR TITLE
client/webserver: use http.Server.Shutdown to stop

### PR DESCRIPTION
**Pull request by chappjc**
Stopping the listeners abruptly kills the server. Instead use
http.Server.Shutdown to close gracefully without interrupting active
connections.

Also fix TestClientMap in webserver.

### CODE REVIEW

File: client/webserver/webserver.go [hunk 1]
```diff
@@ -240,26 +241,34 @@ func (s *WebServer) Run(ctx context.Context) {
 		return
```
Comment (JoeGruffins) [inline client/webserver/webserver.go]: I'm fairly sure that listeners are closed properly when calling srv.Shutdown, so I do not think it is necessary to close them separately. 
https://golang.org/pkg/net/http/#Server.Shutdown
> Shutdown works by first closing all open listeners...

Comment (JoeGruffins) [reply client/webserver/webserver.go]: lol, ignore me, I just woke up

Comment (chappjc) [reply client/webserver/webserver.go]: That's my understanding.  Together with the graceful shutdown procedure of shutdown, that's why I replaced listener.Close with srv.Shutdown.  In some client hacking I noticed that `s.srv.Serve(listener)` would error with something about using a closed connection.  This seems to be the problem with closing the listeners manually and not using the Server's method.

File: client/webserver/webserver_test.go [hunk 2]
```diff
@@ -147,11 +146,33 @@ func (r *TReader) Read(p []byte) (n int, err error) {
 func (r *TReader) Close() error { return nil }
 
 type TConn struct {
-	msg []byte
+	msg       []byte
+	reads     [][]byte      // data for ReadMessage
+	respReady chan []byte   // signal from WriteMessage
+	close     chan struct{} // Close tells ReadMessage to return with error
```
Comment (buck54321) [inline client/webserver/webserver_test.go]: I see this is read in `ReadMessage`, but it doesn't look like it's ever actually closed. Am I reading that right?

Comment (chappjc) [reply client/webserver/webserver_test.go]: Ahhh, I suppose the intent was to put that in `(*TConn).Close`

**Pull request merged**